### ISSUE TITLE
Simplify twisted.internet.ssl by implementing it in terms of startTLS

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4905]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4905|
|Type|enhancement|
|Created|2011-02-22 01:11:26Z|
|Branch|https://github.com/twisted/twisted/tree/simplify-ssl-4905|

Though I'm about to eliminate the need for `Port`, `Connector`, `Client`, and `Server` in `twisted.internet.ssl` ([#4854](https://github.com/twisted/twisted/issues/4854)), I noticed some complexity which can be removed independently from the switch to `twisted.protocols.tls`.

Rather than remove this as part of [[#4854](https://github.com/twisted/twisted/issues/4854)](https://github.com/twisted/twisted/issues/4854), it would be good to remove it first, separately, simplifying the #4854 branch by completely avoiding (I hope) any changes to those four classes.

The simplification also extends to removing some hooks from `twisted/internet/tcp.py`.

The extra complexity is:

  1. The `getHost` and `getPeer` methods of both `Client` and `Server` are all basically the same as the base implementations, except without some useful bug fixes ([#3059](https://github.com/twisted/twisted/issues/3059))
  1. `Port.createInternetSocket` creates an `SSL.Connection` when it could just invoke `startTLS` to do that
  1. `_preMakeConnection` can be removed entirely, in favor of a `startTLS` call in `ssl.Server.__init__`

<details><summary>Searchable metadata</summary>

```
trac-id__4905 4905
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_simplify_ssl_4905 branches/simplify-ssl-4905
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1298337086000000 1298337086000000
changetime__1299624456000000 1299624456000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [30835]) Branching to 'simplify-ssl-4905'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/simplify-ssl-4905)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

0. Less code, hooray!

1. There are no tests for getHost/getPort... But in so far as we're now switching to the base class implementation, I suppose that's OK.

2. In the old code in Server, when is connectionMade called? After TCP accept, or after SSL negotiation is done? In the new code, it's the former, I suspect in the old code it was the latter. At the very least this change should be documented in NEWS, but I'd like to hear an argument about why this shouldn't matter to users. I notice clients already use the after-TCP-but-before-SSL semantics?

3. Why did _socketShutdownMethod change on ssl.Port? Oh, because we're listening with TCP socket now. That being the case, you can probably just remove the line, and use the default inherited from tcp.Port.

4. ssl.Connector also appears to duplicate (again, with a missing feature - service names for ports) a bunch of code in tcp.Connector. As long as you're removing unnecessary code in ssl.py, you should fix that too.


Optional:
A. Remove my name as maintainer in ssl.py docstring.

B. The "future plans" section mention "switching tcp into ssl". Given this patch is *using* this feature, perhaps that line could be removed ;)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [30891]) Address review comments

refs [#4905](https://github.com/twisted/twisted/issues/4905)

  * Remove some future plans which are either done or not plans
  * Remove maintainer marker
  * Remove duplicate `_socketShutdownMethod` on `ssl.Port`
  * Turn `ssl.Connector` into a `tcp.Connector` subclass and share some code

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

> But in so far as we're now switching to the base class implementation, I suppose that's OK

I think so, yea.  Certainly having more test coverage would not be bad, though.  I'll file a ticket for it.

> In the old code in Server, when is connectionMade called?

Before and after this change, `connectionMade` is called when the TCP connection comes up, not when the SSL handshake is complete. :/  Hopefully someday we can fix that, but at least for now there's no change in behavior.

> you can probably just remove the line, and use the default inherited from tcp.Port.

Indeed.

> ssl.Connector also appears to duplicate

Hmmm.  Yea, mostly.  Except it sneaks one extra argument into `__init__`, in the middle.  And instantiates `ssl.Client` with slightly that extra argument (again in the middle), rather than `tcp.Client`.  I think `getDestination` is the only thing that's total duplication.  Of course that's 1/3rd of its methods, so yea, okay.  And it doesn't look like subclassing `tcp.Connector` will introduce any complications.

> Optional: A. Remove my name as maintainer in ssl.py docstring

Done

> The "future plans" section mention "switching tcp into ssl".

Heh.  Indeed.  Done.  I also removed "more options", because I don't know what that means.  Also it's not so much a plan as a philosophy.

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set owner to @exarkun|
|-|-|

Looks good, merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [30895]) Merge simplify-ssl-4905

Author: exarkun
Reviewer: itamar
Fixes: [#4905](https://github.com/twisted/twisted/issues/4905)

Remove some duplicate code between `tcp.py` and `ssl.py` by implementing
SSL transports in terms of `ITLSTransport.startTLS`.
